### ISSUE TITLE
Failing to parse house number correctly

### ISSUE THREAD

**Issue by Randyklose** [openvenues/libpostal#288]
Hey, im trying to parse this address:14635-A Baldwin Park Towne Center, Baldwin Park, CA, USA

and the result i got from libpostal is the following: 
{ 
"house": "14635-a baldwin park towne center",
 "city": "baldwin park",
 "state": "ca",
 "country": "usa" 
}

Is there any indicaiton about why libpostal fails to tokenize the house_number "14635-a"? Why did it added all of it as the "house" token?

**Comment by albarrentine** [comment]
It's probably the "-A", which is relatively rare in house numbers, combined with the rare street name which looks kind of more like a building/venue name. Since numbers like that can also occur in building/venue/company names, the model likely doesn't have enough evidence to predict `house_number` for the first token, so predicts `house`, which makes `house` more likely for the subsequent tokens.

Notice that without the "-A" libpostal gets everything correct:

```
> 14635 Baldwin Park Towne Center, Baldwin Park, CA, USA

Result:

{
  "house_number": "14635",
  "road": "baldwin park towne center",
  "city": "baldwin park",
  "state": "ca",
  "country": "usa"
}
```

Also, even with the "-A", if there were a more common street identifier like "St", "Pl", "Ave", etc. added after the "Baldwin Park Towne Center" it would have enough evidence to revise its previous predictions:

```
> 14635-A Baldwin Park Towne Center St, Baldwin Park, CA, USA

Result:

{
  "house_number": "14635-a",
  "road": "baldwin park towne center st",
  "city": "baldwin park",
  "state": "ca",
  "country": "usa"
}
```

I also checked and that address doesn't exist in either OSM or OpenAddresses in Los Angeles County (which is a government data set). When libpostal gets an address wrong, try looking it up in OSM. If it's not there, you can add it, and libpostal will probably be able to parse that address next time it's trained. It's not an immediate fix, but allows folks to have some input into the process.

For v1.1 it's possible we can add a variable to the model that looks for the first occurrence of a street-like phrase in the case of an initial numeric token, which might help cases like this. I do not have a sense of the release date at the moment and am finishing up the deduping API for libpostal right now as part of a contract.

Generally, as I've said before: libpostal is a machine learning model with 99.45% accuracy. That's a pretty great result, and is similar to the level of accuracy required for commercial viability in OCR machines that read ZIP codes when you send a letter through the post office. At the same time, that means if you parse 1000 addresses, libpostal will get at least one token wrong in 5 of them (haven't broken that down by country, probably less in countries like the US where we have lots of training data). We try to deal with major classes of errors on here, city names that don't exist in our data, etc. but may not be able to fix every single test case in a deterministic way. You have to be ok with that to use libpostal, and it's usually far better than whatever people were doing before (if you have a single country/style of address to deal with, you might be able to design a regex that will work better, but maybe not - I know, I've tried, and I know a thing or two about regex). It's not going to get every address perfectly though, so you have to be able to handle that in your application.

What developers typically like about regex is that when a test case doesn't work as expected, they can simply change some code, see the result, and make it work a little better. In libpostal we have to change data instead of code. This article I think explains the differences between writing ML software vs. normal software: https://medium.com/@karpathy/software-2-0-a64152b37c35, though I'm not personally as much of a fan of neural networks, precisely because feature-based models as used in libpostal allow us to better explain and reason about the results.
